dev/com.ibm.ws.microprofile.reactive.messaging.kafka.adapter.impl/src/com/ibm/ws/microprofile/reactive/messaging/kafka/adapter/impl/KafkaProducerImpl.java

```diff
@@ -21,22 +21,40 @@
 /**
  *
  */
-public class KafkaProducerImpl<K, V> extends AbstractKafkaAdapter<org.apache.kafka.clients.producer.KafkaProducer<K, V>> implements KafkaProducer<K, V> {
+public class KafkaProducerImpl<K, V, T> extends AbstractKafkaAdapter<org.apache.kafka.clients.producer.KafkaProducer<K, T>> implements KafkaProducer<K, V> {
 
     private static final String CLAZZ = KafkaProducerImpl.class.getName();
     private static final Logger LOGGER = Logger.getLogger(CLAZZ);
 
     public KafkaProducerImpl(Map<String, Object> producerConfig) {
-        super(new org.apache.kafka.clients.producer.KafkaProducer<K, V>(producerConfig));
+        super(new org.apache.kafka.clients.producer.KafkaProducer<K, T>(producerConfig));
     }
 
     /** {@inheritDoc} */
     @Override
-    public void send(String topic, V value, Callback callback) {
+    public void send(String configuredTopic, V value, Callback callback, boolean useProducerRecordTopic) {
         if (LOGGER.isLoggable(Level.FINEST)) {
-            LOGGER.logp(Level.FINEST, CLAZZ, "send", "Topic: {0}, Value: {1}", new String[] { topic, value.toString() });
+            LOGGER.logp(Level.FINEST, CLAZZ, "send", "Configured Topic: {0}, Value: {1}, Use ProducerRecord topic: {2}",
+                        new String[] { configuredTopic, value.toString(), Boolean.toString(useProducerRecordTopic) });
+        }
+
+        org.apache.kafka.clients.producer.ProducerRecord<K, T> delegateRecord;
+
+        if (value instanceof org.apache.kafka.clients.producer.ProducerRecord) {
+            org.apache.kafka.clients.producer.ProducerRecord<K, T> userProducerRecord = (org.apache.kafka.clients.producer.ProducerRecord<K, T>) value;
+            org.apache.kafka.common.header.Headers headers = userProducerRecord.headers();
+            K key = userProducerRecord.key();
+            Integer partition = userProducerRecord.partition();
+            Long timestamp = userProducerRecord.timestamp();
+            String userTopic = userProducerRecord.topic();
+            T userValue = userProducerRecord.value();
+
+            String topic = useProducerRecordTopic ? userTopic : configuredTopic;
+
+            delegateRecord = new org.apache.kafka.clients.producer.ProducerRecord<>(topic, partition, timestamp, key, userValue, headers);
+        } else {
+            delegateRecord = new org.apache.kafka.clients.producer.ProducerRecord<>(configuredTopic, (T) value);
         }
-        org.apache.kafka.clients.producer.ProducerRecord<K, V> delegateRecord = new org.apache.kafka.clients.producer.ProducerRecord<>(topic, value);
 
```

Comment:
I don't really like putting this logic in what's meant to be a thin wrapper layer.

I get that if we want to handle Kafka classes directly, then we can only do so so in the adaptor layer, but maybe we could separate the functionality for turning a value into a `ProducerRecord` from the simple wrapping logic.

Reply:
A ProducerRecord is now constructed through the factory so the send method now mirrors the original API again.